From: Nick Mathewson <nickm@torproject.org>
Date: Fri, 26 Jun 2026 11:04:48 -0400
Subject: [PATCH] http: Discard trailers.

Backport of 87e8e44fa774e9677b089b1a5114ee68aefa1636 for libevent 2.1.12.
Regenerated to apply without dependent 2.1.13 HTTP refactor commits.

Fixes GHSA-2gmv-p5m7-98p6 / CVE-2026-63379.
---
--- a/http.c	2020-07-05 08:01:34.000000000 -0400
+++ b/http.c	2026-08-28 08:28:24.853332041 -0400
@@ -181,6 +181,10 @@
 static evutil_socket_t bind_socket(const char *, ev_uint16_t, int reuse);
 static void name_from_addr(struct sockaddr *, ev_socklen_t, char **, char **);
 static struct evhttp_uri *evhttp_uri_parse_authority(char *source_uri);
+static enum message_read_status evhttp_parse_headers_impl_(
+	struct evhttp_request *req,
+	struct evbuffer *buffer,
+	struct evkeyvalq *headers);
 static int evhttp_associate_new_request_with_connection(
 	struct evhttp_connection *evcon);
 static void evhttp_connection_start_detectclose(
@@ -997,8 +1001,10 @@
 evhttp_read_trailer(struct evhttp_connection *evcon, struct evhttp_request *req)
 {
 	struct evbuffer *buf = bufferevent_get_input(evcon->bufev);
+	struct evkeyvalq tmp_headers;
+	TAILQ_INIT(&tmp_headers);
 
-	switch (evhttp_parse_headers_(req, buf)) {
+	switch (evhttp_parse_headers_impl_(req, buf, &tmp_headers)) {
 	case DATA_CORRUPTED:
 	case DATA_TOO_LONG:
 		evhttp_connection_fail_(evcon, EVREQ_HTTP_DATA_TOO_LONG);
@@ -1012,6 +1018,8 @@
 	default:
 		break;
 	}
+
+	evhttp_clear_headers(&tmp_headers);
 }
 
 static void
@@ -2092,14 +2100,17 @@
 	return (0);
 }
 
-enum message_read_status
-evhttp_parse_headers_(struct evhttp_request *req, struct evbuffer* buffer)
+/* As `evhttp_parse_headers_`, but put any headers we find into `headers`. */
+static enum message_read_status
+evhttp_parse_headers_impl_(
+	struct evhttp_request *req,
+	struct evbuffer *buffer,
+	struct evkeyvalq *headers)
 {
 	enum message_read_status errcode = DATA_CORRUPTED;
 	char *line;
 	enum message_read_status status = MORE_DATA_EXPECTED;
 
-	struct evkeyvalq* headers = req->input_headers;
 	size_t len;
 	while ((line = evbuffer_readln(buffer, &len, EVBUFFER_EOL_CRLF))
 	       != NULL) {
@@ -2190,6 +2201,12 @@
 	return (0);
 }
 
+enum message_read_status
+evhttp_parse_headers_(struct evhttp_request *req, struct evbuffer *buffer)
+{
+	return evhttp_parse_headers_impl_(req, buffer, req->input_headers);
+}
+
 static int
 evhttp_method_may_have_body(enum evhttp_cmd_type type)
 {
